nomad/worker.go

```diff
@@ -224,56 +224,24 @@ func (w *Worker) sendAck(evalID, token string, ack bool) {
 	}
 }
 
-// waitForIndex ensures that the local state is at least as fresh
-// as the given index. This is used before starting an evaluation,
-// but also potentially mid-stream. If a Plan fails because of stale
-// state (attempt to allocate to a failed/dead node), we may need
-// to sync our state again and do the planning with more recent data.
-func (w *Worker) waitForIndex(index uint64, timeout time.Duration) error {
-	// XXX: Potential optimization is to set up a watch on the state stores
-	// index table and only unblock via a trigger rather than timing out and
-	// checking.
-
+// snapshotAfter times calls to StateStore.SnapshotAfter which may block.
+func (w *Worker) snapshotAfter(waitIndex uint64, timeout time.Duration) (*state.StateSnapshot, error) {
 	start := time.Now()
-	defer metrics.MeasureSince([]string{"nomad", "worker", "wait_for_index"}, start)
-CHECK:
-	// Get the states current index
-	snapshotIndex, err := w.srv.fsm.State().LatestIndex()
-	if err != nil {
-		return fmt.Errorf("failed to determine state store's index: %v", err)
-	}
-
-	// We only need the FSM state to be as recent as the given index
-	if index <= snapshotIndex {
-		w.backoffReset()
-		return nil
-	}
-
-	// Check if we've reached our limit
-	if time.Now().Sub(start) > timeout {
-		return fmt.Errorf("sync wait timeout reached")
-	}
-
-	// Exponential back off if we haven't yet reached it
-	if w.backoffErr(backoffBaselineFast, backoffLimitFast) {
-		return fmt.Errorf("shutdown while waiting for state sync")
-	}
-	goto CHECK
+	ctx, cancel := context.WithTimeout(w.srv.shutdownCtx, timeout)
+	snap, err := w.srv.fsm.State().SnapshotAfter(ctx, waitIndex)
+	cancel()
+	metrics.MeasureSince([]string{"nomad", "worker", "wait_for_index"}, start)
+	return snap, err
```

Comment:
Wrap this error, if it hits its timeout it will just be context cancelled. `fmt.Errorf("failed to snapshot state at index after %d: %v", waitIndex, err)`

Reply:
Good call. Fixed.